### PR TITLE
Limit the number of concurrent FSMs

### PR DESCRIPTION
**Pull request by urbanserj**

### CODE REVIEW

**Comment by nlsun** [comment]
does this depend on the changes made in the http api PR? if it doesn't, can you remove that commit (you can do that with a rebase -i)

**Comment by urbanserj** [comment]
@nlsun I've tried to rebase my PR. It would have 3 minor and 1 big merge conflicts if I do that.

File: src/spartan_sup.erl [hunk 1]
```diff
@@ -67,9 +58,13 @@ init([]) ->
     ok = localhost_zone_setup(),
 
     %% Systemd Sup intentionally goes last
-    Children = [HandlerSup, ZkRecordServer, ConfigLoaderServer, WatchdogSup],
+    Children = [ZkRecordServer, ConfigLoaderServer, WatchdogSup],
     Children1 = maybe_add_udp_servers(Children),
 
+    sidejob:new_resource(
```
Comment (nlsun) [inline src/spartan_sup.erl]: should this go before maybe_add_udp_servers?

File: src/spartan_handler_fsm.erl [hunk 2]
```diff
@@ -55,9 +55,16 @@
     start_timestamp = undefined :: os:timestamp()
 }).
 
--spec(start_link(from(), binary()) -> {ok, pid()} | ignore | {error, error()}).
-start_link(From, Data) ->
-    gen_fsm:start_link(?MODULE, [From, Data], []).
+-spec(start(from(), binary() | dns_message()) -> {ok, pid()} | ignore | {error, error()}).
```
Comment (nlsun) [inline src/spartan_handler_fsm.erl]: is {error, overload} the only error that can come out of this?

File: src/spartan_udp_server.erl [hunk 3]
```diff
@@ -26,7 +26,8 @@
 %%% API
 %%%===================================================================
 
-do_reply(To = {Pid, IP, Port}, Data) ->
+-spec do_reply(spartan_handler_fsm:from_key(), binary()) -> ok.
```
Comment (GoelDeepak) [inline src/spartan_udp_server.erl]: Can we have same style for spec?

File: src/spartan_sup.erl (hunk 1 above)
Comment (GoelDeepak) [inline src/spartan_sup.erl]: it doesn't matter because we are just setting up

Comment (urbanserj) [inline src/spartan_sup.erl]: sidejob starts separate supervisor even before spartan_sup starts its children.

**Pull request merged**